contentful/apps
# feat: scope token to teamId and add frontend error handling functionality for Vercel config page

## Purpose

The purpose of this PR is to accomplish the following: 
1. Ensure that we are scoping our API communication with Vercel to the provided `teamId` that is associated with the access token. 
2. Communicate all errors/warnings effectively to the user.

## Approach

1. In order to implement the `teamId` scope, I created another method within the `VercelClient` to fetch the "current" token metadata. With this data, if the token metadata contains a `teamId`, this is stored as part of the installation state, and subsequently used as a param when fetching the projects and API paths. If the token does not contain a `teamId`, an error is thrown and appropriately shown to the user. 
2. In order to effectively manage errors on the frontend, I implemented an `errorsReducer`, similar to our current `parametersReducer`. Originally errors were handled on a one-off basis, but this was not scalable, especially with edge cases continually being found. Instead, we can now easily manage the strictly typed states of errors appropriately by using the store as the single source of truth. 

With the above work I created a handful of new hooks that abstract logic from jumbled component code: 
1. `useError` hook returns the appropriate error message based on the current error supplied, and indicates if an error exists at all. 
2. `useFetchAndValidateData` hook returns three methods to effectively handle communication with the token, project, and API path API. 

## Testing steps

Access token section: 
1. Paste an invalid token (i.e '1234') and notice the error message that appears. 
2. Paste a token that is scoped to the entire account (not a team) and notice the error message that appears. 
3. Paste a valid token scoped to a team, and notice no error message. 

Project and API path sections are a bit tricky to test, but you can create a project within Vercel, configure that project on the configuration page, and then delete the project to notice an error message. 

Content type/preview path section:
1. Save an empty preview path, and notice the error message that appears. 
2. Save another preview path that is formatted incorrectly, and notice the error message that appears. 

## Follow-up work

- I have not yet added much nuanced handling of when there are no API paths retrieved (and we instead render a text input). There is a follow-up ticket to capture this work.
